### PR TITLE
Typing consistency 

### PR DESCRIPTION
**Pull request by hamzaelsaawy**
Several changes here related to ensuring consistent return types for `FilterCoefficient` design methods:

-  `Chebyshev`/`Butterworth`/`Elliptic(::T, ...)` now return `ZeroPoleGain{Complex{T}, Complex{T}, T}` instead of cases where the gain was sometimes `Float64`
    - having `ZeroPoleGain{T, Complex{T}, T}` for `Butterworth`/`Chebyshev1` causes some questions/issues with designing a bandpass/stop filter, which returns `ZeroPoleGain{Complex{T}, Complex{T}, T}`, hence I chose `Complex` for roots by default
- `transform_protoype` (and therefore `digitalfilter`) returns `ZeroPoleGains` with the same types as the prototype
- added tests for the above two
- removed `realtype`/`complextype` since they replicate `Base. real`/`Base.complex`
- removed `UnionAll[]` inputs to  `ZeroPoleGain` constructors in `test/filter_conversion.jl`/'filter_design.jl`
- edge case for `PolynomalRatio` with two `Poly`s of different types

I think by far, the most invasive change was making the coefficient arrays in `test/filter_conversion.jl` & `test/filter_design.jl` fit in the 80 character width, which causes diff to freak out, but its the change I most strongly believe in 😐

### CODE REVIEW

**Comment by galenlynch** [comment]
Wow, lots of stuff! Would you mind separating this into smaller PRs, each focused on one logical change (e.g. removing `realtype`/`complextype` as its own PR), to make it easier to evaluate each change in isolation?

**Comment by hamzaelsaawy** [comment]
moved out irrelevant changes to separate branches

File: src/Filters/design.jl [hunk 1]
```diff
@@ -191,9 +201,9 @@ function Elliptic(T::Type, n::Integer, rp::Real, rs::Real)
     # Eq. (65)
     v0 = -im/n*asne(im/εp, k1)
 
-    z = Vector{Complex{T}}(undef, 2*div(n, 2))
-    p = Vector{Complex{T}}(undef, n)
-    gain = one(T)
+    z = Vector{CT}(undef, 2*div(n, 2))
+    p = Vector{CT}(undef, n)
+    gain::T = one(T) # somehow this changes type ...
```
Comment (martinholters) [inline src/Filters/design.jl]: I'd prefer if we could figure out/eliminate where this changes type. My guess is that `ze` and `pole` below are `Complex{Float64}`, no matter what `T` is. Maybe it's sufficient to use `z[2i]`/`p[2i]` (or `p[end]`) below when updating `gain`.

**Comment by galenlynch** [comment]
That's not to say they're not welcome! It's just easier to consider each change in isolation.

File: src/Filters/design.jl [hunk 2]
```diff
@@ -35,8 +37,10 @@ Butterworth(n::Integer) = Butterworth(Float64, n)
 # Chebyshev type I and II prototypes
 #
 
-function chebyshev_poles(T::Type, n::Integer, ε::Real)
-    p = zeros(Complex{T}, n)
+function chebyshev_poles(::Type{T}, n::Integer, ε::Real) where {T<:Real}
+    CT = Complex{T}
+
+    p = zeros(CT, n)
```
Comment (martinholters) [inline src/Filters/design.jl]: This seems to be the only use of `CT`. Maybe not worth introducing the extra variable?

Also in `Chebyshev1` and `Chebyshev2` below.

**Comment by galenlynch** [comment]
> * having `ZeroPoleGain{T, Complex{T}, T}` for `Butterworth`/`Chebyshev1` causes some questions/issues with designing a bandpass/stop filter, which returns `ZeroPoleGain{Complex{T}, Complex{T}, T}`, hence I chose `Complex` for roots by default

Can you expand on this? Both Scipy and Matlab returns real zeros, and complex poles. What is the benefit of returning complex zeros?

**Comment by martinholters** [comment]
Unrelated to this PR, but the coverage submission fails on both 0.7 and 1.0. Anyone any idea what's up with that?

File: src/Filters/design.jl (hunk 1 above)
Comment (hamzaelsaawy) [reply src/Filters/design.jl]: Following through the code, `cde`, `sne`, and `landen` all promote up to their inputs, and `v0 = -im/n*asne(im/εp, k1)` is `Float64` cause of the `im/n` bit, so that propagates through to `pole` and 
therefore `gain`. I think `ze::T`.
I think using `pole` and `ze` improves legibility, and having an explicit type declaration doesn't really hurt anything at all. But we could force `v0::T` instead, but `convert(T, im/n)` doesn't seem all that different than `gain::T`

File: src/Filters/design.jl (hunk 2 above)
Comment (hamzaelsaawy) [reply src/Filters/design.jl]: It was boilerplate I introduced from `Elliptic` and I carried through.
Though I would like to point out that it is trivially optimized out.

File: src/Filters/design.jl [hunk 3]
```diff
@@ -310,23 +321,27 @@ end
 # The Octave implementation was not consulted in creating this code.
 
 # Create a lowpass filter from a lowpass filter prototype
-transform_prototype(ftype::Lowpass, proto::ZeroPoleGain) =
-    ZeroPoleGain(ftype.w * proto.z, ftype.w * proto.p,
+transform_prototype(ftype::Lowpass, proto::ZeroPoleGain{Z, P, K}) where {Z, P, K} =
+    ZeroPoleGain{Z, P, K}(ftype.w * proto.z, ftype.w * proto.p,
```
[the remark this answers is not in the document]
Comment (hamzaelsaawy) [reply src/Filters/design.jl]: So the type parameters will dictate the `Vector{Z}`, etc, type, as per the definition of a constructor. 
Also, the constructor is [tested here](https://github.com/JuliaDSP/DSP.jl/blob/50a03a299ae9d6b850466d3263013a31a3ce7643/test/filter_conversion.jl#L170) since the `convert` call translates to a constructor call which results in a `ZeroPoleGain{Z,P,K}(f.z, f.p, f.k)`, which forces a conversion.

You can test it out with a trivial example via:

```julia
struct A{T}
    x::T
end

A{Int}(2.0)
# A{Int64}(2)
```

Finally the test sets I added in `filter_design.jl:556` onwards explicitly has those tests.

File: src/Filters/design.jl [hunk 4]
```diff
@@ -335,18 +350,20 @@ function transform_prototype(ftype::Highpass, proto::ZeroPoleGain)
 
     abs(real(num) - 1) < np*eps(real(num)) && (num = 1)
     abs(real(den) - 1) < np*eps(real(den)) && (den = 1)
-    ZeroPoleGain(newz, newp, proto.k * real(num)/real(den))
+    ZeroPoleGain(newz, newp, oftype(k, k * real(num)/real(den)))
```
[the remark this answers is not in the document]
Comment (hamzaelsaawy) [reply src/Filters/design.jl]: It would be the same argument in #142, wouldn't it?
We can call an explicit error saying that this filter cannot be represented by an `Int` gain if that helps.

File: src/Filters/design.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (galenlynch) [reply src/Filters/design.jl]: Ah you're totally right! I never knew that about constructors.

File: src/Filters/design.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (galenlynch) [reply src/Filters/design.jl]: No I mean at this line there is a potential `InexactError`:

```julia
ulia> oftype(1, 1 / 2)
ERROR: InexactError: Int64(0.5)
Stacktrace:
 [1] Type at ./float.jl:703 [inlined]
 [2] convert at ./number.jl:7 [inlined]
 [3] oftype(::Int64, ::Float64) at ./essentials.jl:334
 [4] top-level scope at none:0
```

**Comment by hamzaelsaawy** [comment]
Finally, the changes to `Complex` zeros are in anticipation of a change I presented in #258, where I would like to parameterize the roots by the same type, similar to `ControlSystems.jl`.  
I think the only time we have real roots are when they are zero, and to me that special case does not justify the need to have different types for poles and zeros.  
Also, I  am not sure there is a benefit to keeping the roots real only when they are zero, since most other cases would result in complex roots and I am curious what use case relies on real zeros being returned.

Also, it creates an unnecessary difference in `Chebyshev1` vs `Chebyshev2` where the latter returns complex roots, so standardization there was also a motivator

File: src/Filters/design.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (hamzaelsaawy) [reply src/Filters/design.jl]: Im sorry, I misspoke 😅.   
I was trying to say that, similar to the case discussed in #142, if the user inputs a filter with `Int` gain, they must have a preference to that. (Assuming we do default construction to `K::Float64`, which is partially the case since the prototype functions generate a gain for the same "real-type" as the poles).   
So the `InexactError` is a consequence of asking for an integer gain when one cannot maintain the same gain as the prototype.

**Comment by galenlynch** [comment]
> Unrelated to this PR, but the coverage submission fails on both 0.7 and 1.0. Anyone any idea what's up with that?

I must have missed it. The latest build seems all right.

File: src/Filters/design.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (galenlynch) [reply src/Filters/design.jl]: IMO preserving types is great, where possible. However, I think it's better to have a function that does not preserve input types, rather than a function that throws errors or intentionally does not work for some input types. Of course, if there is a compelling reason for why it's necessary in this case, then that would be different. What is the problem that you were experiencing or envisioning that motivated this change? I also don't have a good sense for how often people construct zero pole gain filters with integer datatypes, so maybe I'm just wringing my hands over a very niche corner case?

Maybe you could just add a type bound on the filter object to limit these functions to gain types that are closed under division?

File: src/Filters/design.jl (hunk 2 above)
Comment (galenlynch) [reply src/Filters/design.jl]: It definitely doesn't affect the generated code, but it makes the code more indirect and therefore harder to follow.

File: src/Filters/design.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (hamzaelsaawy) [reply src/Filters/design.jl]: It was more of a motivation to preserve typing. Similar to how Julia errors for `Int(3.4)`, the assumption I have is the user explicitly creates a filter with the types they want. And if they create a filter with an integer gain, then they must have explicitly wanted one.

But I definitely think that some people want to have integer-only filters

[the remark this answers is not in the document]
Comment (galenlynch) [reply src/Filters/design.jl]: TBH I doubt these functions would work with integer data types before this PR. `newz[i] = ftype.w / z[i]` would also be problematic.

[the remark this answers is not in the document]
Comment (hamzaelsaawy) [reply src/Filters/design.jl]: it might be worthwhile to figure out what people typically expect when they construct a filter and the use cases they have. I am curious if someone would create an integer filter and then modify that, or do all the design work with floats/fixed point and then convert to integers at the end

[the remark this answers is not in the document]
Comment (galenlynch) [reply src/Filters/design.jl]: Since it's clear that this function would not have worked with integers in the first place, I'm ok with it still not working with integers, for now.

File: src/Filters/design.jl (hunk 1 above)
Comment (galenlynch) [reply src/Filters/design.jl]: It makes a substantial difference in the generated code, which you can see even by looking at `@code_warntype`. Stopping the unwanted type promotion at the source, with `convert(T, im/n)` leads to most intermediate variables being of type `T`, while using type annotation with `gain::T` leads most of the intermediate variable to being `Float64`. I think it's cleaner to only use type annotation where absolutely necessary, so I would also vote for using `convert(T, im/n)` over `gain::T`.

Comment (hamzaelsaawy) [reply src/Filters/design.jl]: I did not know that.
I changed it to `im/convert(T, n)`.

**Comment by galenlynch** [comment]
Thanks!
